components/timezone/src/ids.rs

```diff
@@ -0,0 +1,371 @@
+                if candidate.index() == trie_value.index() && candidate.is_canonical() {
+                    // Success! Found what we were looking for.
+                    break;
+                }
+            }
+            // Now check for children of the current node.
+            let mut sub_cursor = cursor.clone();
+            if let Some(probe_result) = sub_cursor.probe(index) {
+                // Found a child. Add the current byte edge to the string.
+                if !probe_result.byte.is_ascii() {
+                    debug_assert!(false, "non-ASCII probe byte: {}", probe_result.byte);
+                    return None;
+                }
+                // Safety: the byte being added is ASCII as guarded above
+                unsafe { string.to_mut().as_mut_vec().push(probe_result.byte) };
+                // Add the child to the stack, and also add back the current
+                // node if there are more siblings to visit.
+                if index + 1 < probe_result.total_siblings as usize {
+                    stack.push((cursor, index + 1, suffix_len));
+                    stack.push((sub_cursor, 0, 1));
+                } else {
+                    stack.push((sub_cursor, 0, suffix_len + 1));
+                }
+            } else {
+                // No more children. Pop this node's bytes from the string.
+                for _ in 0..suffix_len {
+                    // Safety: we check that the bytes being removed are ASCII
+                    let removed_byte = unsafe { string.to_mut().as_mut_vec().pop() };
+                    if let Some(removed_byte) = removed_byte {
+                        if !removed_byte.is_ascii() {
+                            debug_assert!(false, "non-ASCII removed byte: {removed_byte}");
+                            return None;
+                        }
+                    } else {
+                        debug_assert!(false, "could not remove another byte");
+                        return None;
+                    }
+                }
+            }
+        }
+        Some(NormalizedIana { string, bcp47_id })
+    }
+
+    /// Returns the canonical, normalized IANA ID of the given BCP-47 ID.
+    ///
+    /// Only use this function if you don't have the IANA ID. [`Self::canonicalize_iana()`]
+    /// is much faster in the common case.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_timezone::TimeZoneBcp47Id;
+    /// use icu_timezone::TimeZoneIdMapper;
+    /// use tinystr::tinystr;
+    ///
+    /// let mapper = TimeZoneIdMapper::new();
+    /// let mapper = mapper.as_borrowed();
+    ///
+    /// let canonicalized = mapper.canonicalize_iana("Asia/Calcutta").unwrap();
+    ///
+    /// assert_eq!(canonicalized.string, "Asia/Kolkata");
+    /// assert_eq!(
+    ///     canonicalized.bcp47_id,
+    ///     TimeZoneBcp47Id(tinystr!(8, "inccu"))
+    /// );
+    /// ```
+    pub fn bcp47_to_iana_search(&self, bcp47_id: TimeZoneBcp47Id) -> Option<String> {
```

Comment:
thought: why don't we make this return `NormalizedIana<'static>` as well? We do guarantee to return normalized IANA ids, and have the bcp id.

Reply:
Obsolete since `NormalizedIana` is no more